iroh-pkarr-node-discovery/src/lib.rs

```diff
@@ -76,83 +93,245 @@ impl<'a> Builder<'a> {
+                            // not working is due to a network issue, and if the network changes
+                            // the task will be restarted anyway.
+                            //
+                            // Being unable to publish to the DHT is something that is expected
+                            // to happen from time to time, so this does not warrant a error log.
+                            tracing::warn!("pkarr publish error: {}", e);
+                        }
+                    }
+                }
+            };
+            // publish to the relay if enabled
+            let relay_publish = async {
+                if let Some(url) = this.0.relay_url.as_ref() {
+                    tracing::info!("publishing to relay: {}", url);
+                    match this.0.pkarr.relay_put(url, &signed_packet).await {
+                        Ok(_) => {
+                            tracing::debug!("pkarr publish to relay success");
+                        }
+                        Err(e) => {
+                            tracing::warn!("pkarr publish to relay error: {}", e);
+                        }
+                    }
+                }
+            };
+            // do both at the same time
+            tokio::join!(relay_publish, dht_publish);
+            tokio::time::sleep(REPUBLISH_DELAY - INITIAL_PUBLISH_DELAY).await;
+        }
+    }
+
+    async fn resolve_relay(
+        &self,
+        pkarr_public_key: PublicKey,
+        co: &Co<anyhow::Result<DiscoveryItem>>,
+    ) {
+        let Some(url) = &self.0.relay_url else {
+            return;
+        };
+        tracing::info!("resolving {} from relay {}", pkarr_public_key.to_z32(), url);
+        let response = self.0.pkarr.relay_get(url, pkarr_public_key).await;
+        match response {
+            Ok(Some(signed_packet)) => {
+                if let Ok(node_info) = NodeInfo::from_pkarr_signed_packet(&signed_packet) {
+                    let addr_info = node_info.into();
+                    tracing::info!("discovered node info from relay {:?}", addr_info);
+                    co.yield_(Ok(DiscoveryItem {
+                        provenance: "relay",
+                        last_updated: None,
+                        addr_info,
+                    }))
+                    .await;
+                } else {
+                    tracing::debug!("failed to parse signed packet as node info");
+                }
+            }
+            Ok(None) => {
+                tracing::debug!("no signed packet found in relay");
+            }
+            Err(e) => {
+                tracing::debug!("failed to get signed packet from relay: {}", e);
+                co.yield_(Err(e.into())).await;
+            }
+        }
+    }
+
+    /// Resolve a node id from the DHT.
+    async fn resolve_dht(
+        &self,
+        pkarr_public_key: PublicKey,
+        co: &Co<anyhow::Result<DiscoveryItem>>,
+    ) {
+        if !self.0.dht {
+            return;
```

Comment:
why not yield an error?

Reply:
but if both are disabled you end up with a generator that will yield nothing without error, which seems unfortunate

Reply:
I think it is impossible to get this. There is something about this in the builder.

Reply:
ok